packages/cosmwasm-stargate/src/instantiate2.ts

```diff
@@ -0,0 +1,69 @@
+import { Sha256, sha256 } from "@cosmjs/crypto";
+import { fromBech32, toAscii, toBech32, toUtf8 } from "@cosmjs/encoding";
+import { Uint64 } from "@cosmjs/math";
+import { assert } from "@cosmjs/utils";
+
+/**
+ * The "Basic Address" Hash from
+ * https://github.com/cosmos/cosmos-sdk/blob/v0.45.8/docs/architecture/adr-028-public-key-addresses.md
+ */
+function hash(type: string, key: Uint8Array): Uint8Array {
+  return new Sha256(sha256(toAscii(type))).update(key).digest();
+}
+
+/**
+ * Takes an integer [0, 2**64-1] and returns a one-byte encoding of it.
+ */
+function toUint64(int: number): Uint8Array {
+  return Uint64.fromNumber(int).toBytesBigEndian();
+}
+
+/**
+ * Private function to export test vector data for https://github.com/cosmos/cosmjs/pull/1253.
+ * Do not use in production code.
+ */
+// eslint-disable-next-line @typescript-eslint/naming-convention
+export function _instantiate2AddressIntermediate(
+  checksum: Uint8Array,
+  creator: string,
+  salt: Uint8Array,
+  msg: string | null,
+  prefix: string,
+): { key: Uint8Array; addressData: Uint8Array; address: string } {
+  assert(checksum.length === 32);
+  const creatorData = fromBech32(creator).data;
+
+  const msgData = typeof msg === "string" ? toUtf8(msg) : new Uint8Array();
+
+  // Validate inputs
+  if (salt.length < 1 || salt.length > 64) throw new Error("Salt must be between 1 and 64 bytes");
+
+  const key = new Uint8Array([
+    ...toAscii("wasm"),
+    0x00,
+    ...toUint64(checksum.length),
+    ...checksum,
+    ...toUint64(creatorData.length),
+    ...creatorData,
+    ...toUint64(salt.length),
+    ...salt,
+    ...toUint64(msgData.length),
+    ...msgData,
+  ]);
+  const addressData = hash("module", key);
+  const address = toBech32(prefix, addressData);
+  return { key, addressData, address };
+}
+
+/**
+ * Predictable address generation for the MsgInstantiateContract2
+ * introduced with wasmd 0.29.
+ */
+export function instantiate2Address(
+  checksum: Uint8Array,
+  creator: string,
+  salt: Uint8Array,
+  prefix: string,
+): string {
+  return _instantiate2AddressIntermediate(checksum, creator, salt, null, prefix).address;
```

Comment:
Why msg is always set to null when instantiate2 while Wasmd v0.31.x include msgs when creating address?

Reply:
See https://medium.com/cosmwasm/dev-note-3-limitations-of-instantiate2-and-how-to-deal-with-them-a3f946874230

Reply:
I added https://github.com/cosmos/cosmjs/pull/1398 for future readers

Reply:
I got it. 
However, according to the link you provided, it seems that fix_msg should be null, not that instantiate msg should be null and current code is making instantiate msg to null.

Reply:
To be honest, I am not familiar with wasmd, so I'm not sure how to use instantiate2. 
If you have a chance, could you make a test code that compares the address obtained with instantiate2Address and the address of the contract instantiated with instantiate2 in the script/wasmd chain?

Reply:
Yeah, it's still early for instantiate2 and not much resources are out there. Feel free to subscribe to https://github.com/cosmos/cosmjs/issues/1399 to see
- Pre-calculation
- Actual instantiation
in CosmJS. The tests for this ticket will make a few things clearer I guess.

Reply:
Oh, nice! I **strongly** advise for removing the msg input field there. Nobody should need to use it.